### PR TITLE
[Modbusino] - simple modbus rtu library

### PR DESCRIPTION
**Pull request by kmihaylov**
This PR introduces Modbusino. Component and sample with modifications for Sming.

The most important modification is the replacement of Serial.flush() with Serial.clear().
I removed some code that was spamming the line in case of MB errors.
I also added some minimal delays for the interface IC to ensure proper line detection when transmitting data.

Four environment variables exist for the component and one config variable is for the application (the modbus address).

### CODE REVIEW

File: Sming/Libraries/modbusino/README.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
+modbusino RTU Library (modbus slave)
+====================================
+
+The author's project is at https://github.com/stephane/modbusino
+
+modbusino is lightweight RTU modbus slave library that supports 'read holding registers' and 'write multiple registers' functions.
+
+Four environment variables deifne the RS485 interface hardware pins and pre- and post-transmission delay.
```
Comment (slaff) [inline Sming/Libraries/modbusino/README.rst]: Replace `deifne` with `define`. Make sure to run a spell checker.

Comment (kmihaylov) [reply Sming/Libraries/modbusino/README.rst]: To be honest my Eclipse's spell checker doesn't work. I'm investigating.

**Comment by mikee47** [comment]
I've had a longer think about this and I think to do this properly involves integrating master and slave functionality in the same library as there's a lot of common code involved. Rewriting the modbusino library as I suggested above is probably not the best way forward so happy with this PR as-is, minor points notwithstanding.

File: Sming/Libraries/modbusino/samples/generic/README.rst [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Modbusino RTU generic sample
+============================
+
```
[the remark this answers is not in the document]
Comment (kmihaylov) [reply Sming/Libraries/modbusino/samples/generic/README.rst]: Do you mean to set the default style to bash (first line)?

File: Sming/Libraries/modbusino/README.rst [hunk 3]
```diff
@@ -0,0 +1,27 @@
+modbusino RTU Library (modbus slave)
+====================================
+
+The author's project is at https://github.com/stephane/modbusino
```
[the remark this answers is not in the document]
Comment (kmihaylov) [reply Sming/Libraries/modbusino/README.rst]: Ok, I'm removing it.

File: Sming/Libraries/modbusino/README.rst [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+modbusino is lightweight RTU modbus slave library that supports 'read holding registers' and 'write multiple registers' functions.
+
+Four environment variables define the RS485 interface hardware pins and pre- and post-transmission delay.
```
[the remark this answers is not in the document]
Comment (kmihaylov) [reply Sming/Libraries/modbusino/README.rst]: OK!

File: Sming/Libraries/modbusino/README.rst [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+
+.. envvar:: MB_PRE_TX_DELAY
+
+   With :envvar:`MB_PRE_TX_DELAY` (in microseconds) the delay between changing the RE pin level and sending data can be set. Used to ensure proper line detection by the master device.
+
+
+.. envvar:: MB_POST_TX_DELAY
+
+   :envvar:`MB_POST_TX_DELAY` is the time in microseconds between the end of data transmission and releasing the line.
```
[the remark this answers is not in the document]
Comment (kmihaylov) [reply Sming/Libraries/modbusino/README.rst]: OK!

**Comment by kmihaylov** [comment]
Thank you for your corrections, these README files would be good templates.

Your opinion about the polling changed my mind about how hard would this to be implemented. If you're not in a hurry I may try these changes (in few weeks probably)?

My idea was to include some stuff while your library is getting prepared. I thought it supports both master and client? Maybe I'm in a mistake!

**Comment by mikee47** [comment]
When transmitting, you can express the pre/post delays in terms of characters
then use the 'transmit complete' serial callback to set RE inactive again.

This means you don't need to wait around for the transmission to complete,
and you no longer require pre/post delay variables since the delays are now tied to the baud rate.

You could also use this approach in ModbusMaster.

```
void ModbusinoSlave::setup(long baud)
{
    Serial.begin(baud);

    // When any transmission completes, set RE line inactive
    Serial.onTransmitComplete(
        [](HardwareSerial &) { digitalWrite(RS485_RE_PIN, !RS485_TX_LEVEL); });
}

static void send_msg(uint8_t *msg, uint8_t msg_length)
{
    uint16_t crc = crc16(msg, msg_length);

    msg[msg_length++] = crc >> 8;
    msg[msg_length++] = crc & 0x00FF;

    // Set RE active, add a one-character delay before and after message
    digitalWrite(RS485_RE_PIN, RS485_TX_LEVEL);
    constexpr uint8_t NUL{0};
    Serial.write(NUL);
    Serial.write(msg, msg_length);
    Serial.write(NUL);

    // RE is set inactive by Serial transmit-complete callback
}

```

**Comment by mikee47** [comment]
> My idea was to include some stuff while your library is getting prepared. I thought it supports both master and client?

https://github.com/SmingHub/Sming/issues/1991#issuecomment-563244700

**Comment by kmihaylov** [comment]
OK let me give it a try and I'll try to make a commit today.

**Comment by kmihaylov** [comment]
I think I'm getting closer to removing loops and use callbacks instead. The current state of the code successfully receives write multiple registers (0x10) and replies. Probably I'm doing something terrible with the dataPtr pointer, because the result isn't filled.

On the master device I'm getting timeout errors, but they're probably due some bug in the master. They appeared when I changed the delays with NUL's method. It seems that the loops in the master are going to me removed too... If I get a little bit more efficient.

**Comment by mikee47** [comment]
It's the pre-delay causing the timeout, master interprets it as first character of message. What happens if you just remove the first NUL after asserting TX_EN ?

**Comment by slaff** [comment]
@kmihaylov What is the state of this PR?

**Comment by kmihaylov** [comment]
@slaff, there was still a problem with the timing. I'll try to finish the task in the first half of the week.

**Comment by kmihaylov** [comment]
I found problem both in modbusino and in ModbusMaster. It is related to ```onTransmitComplete``` callback in modbusino. In ModbusMaster it isn't used, but is similar -> the introduced (posttransmission) delay might be not enough and the RE pin might be switched before the actual serial transmission is finished.

It's just the program that goes faster than the UART.

#1645

**Comment by kmihaylov** [comment]
The latest update of ```modbusino``` library introduces the following updates (I may forget something):
1. Remove NUL character from the beginning of a message (suspected to trigger TX FIFO empty interrupt that would release the TX line).
2. Add a NUL characted to the end of the message.
3. The length of the received message is now not strictly compared (==), but 'at least' (>=). The message length is calculated from the information contained in the message frame, and not by the length of the UART hardware.
4. The RX buffer wasn't cleared on successful message. This is now fixed. In the past, under certain circumstances proper work of the driver was observed due to the introduced delays, that would trigger a timeout and would clear the RX buffer, which is not how the program should work.

Unfortunately too many errors somehow led to proper work of the previous versions. I hope that this release is much more stable and production-ready.

**Comment by slaff** [comment]
> Unfortunately too many errors somehow led to proper work of the previous versions. I hope that this release is much more stable and production-ready.

@kmihaylov does it mean that this PR is ready for review and merging?

**Comment by kmihaylov** [comment]
Yes.

**Comment by slaff** [comment]
> Yes.

@mikee47 @kpishere Can you review this PR and test it with a real device?

**Comment by kmihaylov** [comment]
If boards are needed maybe I can help.

File: Sming/Libraries/modbusino/samples/generic/app/application.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#include <SmingCore.h>
+#include <Debug.h>
+#include <Modbusino.h>
+
+#define ARRLEN 3
+uint16_t mbDataArray[ARRLEN] = {0,0,0};
+ModbusinoSlave MbSlave(MB_SLAVE_ADDR, mbDataArray, ARRLEN);
```
Comment (slaff) [inline Sming/Libraries/modbusino/samples/generic/app/application.cpp:7]: Rename `MbSlave` to `mbSlave` to match the [coding style](https://sming.readthedocs.io/en/latest/contribute/coding-style.html).

File: Sming/Libraries/modbusino/samples/generic/app/application.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#include <SmingCore.h>
```
Comment (slaff) [inline Sming/Libraries/modbusino/samples/generic/app/application.cpp:1]: Add a new empty file named '.cs' inside of this folder 'Sming/Libraries/modbusino/samples/generic/app/application.cpp' and run the following commands to apply the coding style rules:
```
cd $SMING_HOME # change this to %SMING_HOME% if you are using Windows
make cs
``` 
Make sure to include the new .cs file in this PR.

**Comment by slaff** [comment]
@mikee47 Any comments here?

File: Sming/Libraries/modbusino/samples/generic/README.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+.. highlight:: bash
+
+The sample code provides three element uint16_t array used for the modbus slave registers. These values are printed each second through UART1.
+mbSlaveLoopTimer checks for new data each 10 milliseconds (Only checks for new data. Serial communication is handled by hardware.)
```
Comment (mikee47) [inline Sming/Libraries/modbusino/samples/generic/README.rst:7]: Needs updating to reflect actual operation i.e. register values are printed only when a valid WriteMultipleRegisters command is received, no timer.

File: Sming/Libraries/modbusino/samples/generic/README.rst [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+Several environment variables (envvar:`RS485_RE_PIN` :envvar:`RS485_TX_LEVEL` :envvar:`MB_PRE_TX_DELAY` :envvar:`MB_POST_TX_DELAY`) can be used for configuration.
```
Comment (mikee47) [inline Sming/Libraries/modbusino/samples/generic/README.rst:9]: `MB_PRE_TX_DELAY` and `MB_POST_TX_DELAY` don't exist!

File: Sming/Libraries/modbusino/component.mk [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+RS485_TX_LEVEL ?= HIGH
+COMPONENT_CXXFLAGS += -DRS485_TX_LEVEL=$(RS485_TX_LEVEL)
+
+COMPONENT_VARS += MB_PRE_TX_DELAY
+MB_PRE_TX_DELAY ?= 100
+COMPONENT_CXXFLAGS += -DMB_PRE_TX_DELAY=$(MB_PRE_TX_DELAY)
+
+COMPONENT_VARS += MB_POST_TX_DELAY
+MB_POST_TX_DELAY ?= 1000
+COMPONENT_CXXFLAGS += -DMB_POST_TX_DELAY=$(MB_POST_TX_DELAY)
```
Comment (mikee47) [inline Sming/Libraries/modbusino/component.mk:19]: Not used - remove

File: Sming/Libraries/modbusino/README.rst [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+   Default: 100 (microseconds).
+
+   Time to wait after RE pin level change, before writing serial data. Used to ensure proper line detection by the master device.
+
+
+.. envvar:: MB_POST_TX_DELAY
+
+   Default: 1000 (microseconds).
+
+   Delay after data transmission, before RE pin level change.
```
Comment (mikee47) [inline Sming/Libraries/modbusino/README.rst:37]: Not used - remove

**Comment by kmihaylov** [comment]
I'll fix these once I get back to the city, because I'm currently locked out. BTW I'm using 115200, probably that is why I didn't notice the delays you have observed. Maybe I can switch to 9600 and try the interrupt.

**Comment by mikee47** [comment]
It's not actually related to baud rate. Below, the blue trace is _RE_ and red trace is slave response. I see either this:

image

And sometimes this:

image

The delay is caused by the `mbPrint()` callback - commenting it out eliminates the delay. This is because on these occasions the RECEIVE handler is getting called before transmitComplete.

If the master tries to send another command within this 5ms period then it'll get lost, otherwise it may not be an issue.

**Comment by kmihaylov** [comment]
> I note in modbusinoSlave.cpp you've got `uartCallback` commented-out in (lines 66+ and 96+). If you were having trouble getting this to work note that `uartCallback` is invoked in interrupt context so line 66 would require `IRAM_ATTR` adding to the function.

Sure I have. How could I forget `IRAM_ATTR` ! Just did as you propose, however the interrupt did not fire. Before I had some behavior like the interrupt is called after the transmission of the first NUL character (when I didn't use `IRAM_ATTR`). Should I force enable interrupts?

**Comment by mikee47** [comment]
> ... however the interrupt did not fire.

Because the call to `Serial.onDataReceived` resets the interrupt callback. For this to work you'd need to handle receives in your ISR as well. Operation seems good enough without this so I'd suggest just removing the commented-out lines.

> Before I had some behavior like the interrupt is called after the transmission of the first NUL character (when I didn't use `IRAM_ATTR`). Should I force enable interrupts?

Interrupt occurs when the transmit FIFO becomes empty. When idle, that will (probably) happen as soon as you write the first character because it gets moved (almost) immediately from the FIFO into the output shift register. A simple way to mitigate this is to set a flag when the message has been written; ignore any TX_EMPTY interrupt unless that flag has been set.

**Comment by kmihaylov** [comment]
@mikee47 @slaff 
I removed the unused setUartCallback function. Also I added a note to the component's readme describing the RE delay. I understand that better is to replace onTransmitComplete with ISR, so maybe the next update would do so...
Meanwhile I'm sticking with Mike's recommendation to leave it this way.

**Comment by slaff** [comment]
> Meanwhile I'm sticking with Mike's recommendation to leave it this way.

@kmihaylov If the PR is ready in the current form I will schedule it for merging.

**Comment by kmihaylov** [comment]
Let me compile and flash it just to be sure and I'll write if everything works in 15 minutes.

**Comment by kmihaylov** [comment]
Yes, it works as expected. I do not observe RE delays on the default conditions for ModbusMaster and modbusino, as they exist now in the PR (115200). I remind that Mike is using another testing approach.

**Comment by slaff** [comment]
> I remind that Mike is using another testing approach.

@mikee47 Do you want to do a final test before I merge this PR?

**Comment by mikee47** [comment]
@slaff Looks fine, happy to merge.

**Pull request merged**